src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -647,6 +670,48 @@ public static Guid GenerateGuidForType(Type type)
             return type.GUID;
         }
 
+        /// <summary>
+        /// This method generates a PROGID for the specified type. If the type has
+        /// a PROGID in the metadata then it is returned otherwise a stable PROGID
+        /// is generated based on the fully qualified name of the type.
+        /// </summary>
+        public static string GenerateProgIdForType(Type type)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+            if (type.IsImport)
+            {
+                throw new ArgumentException(SR.Argument_TypeMustNotBeComImport, nameof(type));
+            }
+            if (type.IsGenericType)
+            {
+                throw new ArgumentException(SR.Argument_NeedNonGenericType, nameof(type));
+            }
+
+            IList<CustomAttributeData> cas = CustomAttributeData.GetCustomAttributes(type);
+            for (int i = 0; i < cas.Count; i++)
+            {
+                if (cas[i].Constructor.DeclaringType == typeof(ProgIdAttribute))
+                {
+                    // Retrieve the PROGID string from the ProgIdAttribute.
+                    IList<CustomAttributeTypedArgument> caConstructorArgs = cas[i].ConstructorArguments;
```

Comment:
`cas[i]` is virtual call. Cache it in a local?

Reply:
(Alternatively, this can be `foreach` loop.)

Reply:
> (Alternatively, this can be foreach loop.)

Wouldn't that still end up being two virtual calls for each item?

Reply:
Oh, nevermind, I thought you were referring to the overall loop and eliminating the cas.Count on each iteration.